dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/preheat/supplier/UniqueAttributeSupplierTest.java

```diff
@@ -0,0 +1,274 @@
+
+/**
+ * @author Enrico Colasante
+ */
+@ExtendWith( MockitoExtension.class )
+class UniqueAttributeSupplierTest extends DhisConvenienceTest
+{
+
+    private static final String UNIQUE_VALUE = "unique value";
+
+    private static final String TEI_UID = "TEI UID";
+
+    private static final String ANOTHER_TEI_UID = "ANOTHER_TEI_UID";
+
+    @InjectMocks
+    private UniqueAttributesSupplier supplier;
+
+    @Mock
+    private TrackedEntityAttributeService trackedEntityAttributeService;
+
+    @Mock
+    private TrackedEntityAttributeValueService trackedEntityAttributeValueService;
+
+    private TrackerImportParams params;
+
+    private TrackedEntityAttribute uniqueAttribute;
+
+    private TrackedEntityInstance tei;
+
+    private ProgramInstance programInstance;
+
+    private TrackedEntityAttributeValue trackedEntityAttributeValue;
+
+    @BeforeEach
+    public void setUp()
+    {
+        params = TrackerImportParams.builder().build();
+        uniqueAttribute = createTrackedEntityAttribute( 'A', ValueType.TEXT );
+        OrganisationUnit orgUnit = createOrganisationUnit( 'A' );
+        Program program = createProgram( 'A' );
+        Attribute attribute = createAttribute( 'A' );
+        AttributeValue attributeValue = createAttributeValue( attribute, UNIQUE_VALUE );
+        tei = createTrackedEntityInstance( 'A', orgUnit );
+        tei.setUid( TEI_UID );
+        tei.setAttributeValues( Collections.singleton( attributeValue ) );
+        programInstance = createProgramInstance( program, tei, orgUnit );
+        programInstance.setAttributeValues( Collections.singleton( attributeValue ) );
+        trackedEntityAttributeValue = createTrackedEntityAttributeValue( 'A', tei, uniqueAttribute );
+    }
+
+    @Test
+    void verifySupplierWhenNoUniqueAttributeIsPresentInTheSystem()
+    {
+        // given
+        TrackerPreheat preheat = new TrackerPreheat();
+        when( trackedEntityAttributeService.getAllUniqueTrackedEntityAttributes() )
+            .thenReturn( Collections.emptyList() );
+
+        // when
+        this.supplier.preheatAdd( params, preheat );
+
+        // then
+        assertThat( preheat.getUniqueAttributeValues(), hasSize( 0 ) );
+    }
+
+    @Test
+    void verifySupplierWhenTEIAndEnrollmentHaveTheSameUniqueAttribute()
+    {
+        // given
+        TrackerPreheat preheat = new TrackerPreheat();
+        when( trackedEntityAttributeService.getAllUniqueTrackedEntityAttributes() )
+            .thenReturn( Collections.singletonList( uniqueAttribute ) );
+        TrackerImportParams importParams = TrackerImportParams.builder()
+            .trackedEntities( Collections.singletonList( trackedEntity() ) )
+            .enrollments( Collections.singletonList( enrollment( TEI_UID ) ) )
+            .build();
+
+        // when
+        this.supplier.preheatAdd( importParams, preheat );
+
+        // then
+        assertThat( preheat.getUniqueAttributeValues(), hasSize( 0 ) );
```

Comment:
Why does this one return 0 unique attributes, and the test below 2 even both say ...HaveTheSameUniqueAttribute?

Reply:
I have changed the test names to differentiate the ones that found unique attributes and the ones that do not find any.
It is not easy to explain everything in the name though.

We are going to have a validation error if we are using the same value for the same unique attributes for 2 different TEIs or enrollments in different TEIs, and we are NOT going to have a validation error if we have the same value for the same unique attribute for a TEI and an enrollment in this TEI, or if the TEI in the payload is the same as the TEI in the DB (in this case it is just an update)

Reply:
uf 😅 yes, that is hard to put in a method name 😂 thank you for explaining!